src/main/java/org/terasology/logic/health/HealthAuthoritySystem.java

```diff
@@ -119,13 +127,36 @@ private int regenerateHealth(HealthComponent health, int healAmount) {
         return newHeal;
     }
 
-    private void checkHealed(EntityRef entity, HealthComponent health, int healAmount) {
+    private void checkRegenerated(EntityRef entity, HealthComponent health, int healAmount) {
         if (healAmount > 0) {
-            checkHeal(entity, healAmount, entity);
+            BeforeRegenEvent beforeRegen = entity.send(new BeforeRegenEvent(healAmount, entity));
+            if (!beforeRegen.isConsumed()) {
+                int modifiedAmount = TeraMath.floorToInt(beforeRegen.getResultValue());
+                if (modifiedAmount > 0) {
+                    doRegenerate(entity, modifiedAmount, entity);
+                } else if (modifiedAmount < 0) {
+                    doDamage(entity, -modifiedAmount, EngineDamageTypes.HEALING.get(), entity, EntityRef.NULL);
+                }
+            }
             entity.saveComponent(health);
         }
     }
 
+    private void doRegenerate(EntityRef entity, int healAmount, EntityRef instigator) {
+        HealthComponent health = entity.getComponent(HealthComponent.class);
+        if (health != null) {
```

Comment:
These (4 lines) can be taken out as a helper function for both Regen and Restore. Will refactor this after also adding restore system.